gk/fib.c

```diff
@@ -0,0 +1,1440 @@
+				"gk: failed to allocate the IPv6 FIB table at %s\n",
+				__func__);
+			ret = -1;
+			goto free_lpm6;
+		}
+	}
+
+	ret = init_fib_tbl(gk_conf);
+	if (ret < 0) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"gk: failed to initialize the FIB table at %s\n",
+			__func__);
+		goto free_lpm_tbl6;
+	}
+
+	ret = 0;
+	goto out;
+
+free_lpm_tbl6:
+	if (!ipv6_configured(gk_conf->net))
+		goto free_lpm_tbl;
+
+	rte_free(ltbl->fib_tbl6);
+	ltbl->fib_tbl6 = NULL;
+
+free_lpm6:
+	destroy_ipv6_lpm(ltbl->lpm6);
+	ltbl->lpm6 = NULL;
+
+free_lpm_tbl:
+	if (!ipv4_configured(gk_conf->net))
+		goto out;
+
+	rte_free(ltbl->fib_tbl);
+	ltbl->fib_tbl = NULL;
+
+free_lpm:
+	destroy_ipv4_lpm(ltbl->lpm);
+	ltbl->lpm = NULL;
+
+out:
+	return ret;
+}
+
+static int
+notify_gk_instance(struct gk_fib *fib, struct gk_instance *instance)
+{
+	int ret;
+	struct mailbox *mb = &instance->mb;
+	struct gk_cmd_entry *entry = mb_alloc_entry(mb);
+	if (entry == NULL) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"gk: failed to allocate a `struct gk_cmd_entry` entry at %s\n",
+			__func__);
+		return -1;
+	}
+
+	entry->op = GK_SYNCH_WITH_LPM;
+	entry->u.fib = fib;
+
+	ret = mb_send_entry(mb, entry);
+	if (ret < 0) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"gk: failed to send a `struct gk_cmd_entry` entry at %s\n",
+			__func__);
+		return -1;
+	}
+
+	return 0;
+}
+
+/*
+ * XXX What we are doing here is analogous to RCU's synchronize_rcu(),
+ * what suggests that we may be able to profit from RCU. But we are going
+ * to postpone that until we have a better case to bring RCU to Gatekeeper.
+ */
+static void
+synchronize_gk_instances(struct gk_fib *fib, struct gk_config *gk_conf)
+{
+	int i, loop;
+	int num_succ_notified_inst = 0;
+	bool is_succ_notified[gk_conf->num_lcores];
+
+	/* The maximum number of times to try to notify the GK instances. */
+	const int MAX_NUM_NOTIFY_TRY = 3;
+
+	rte_atomic16_init(&fib->num_updated_instances);
+
+	memset(is_succ_notified, false, sizeof(is_succ_notified));
+
+	for (loop = 0; loop < MAX_NUM_NOTIFY_TRY; loop++) {
+		/* Send the FIB entry to the GK mailboxes. */
+		for (i = 0; i < gk_conf->num_lcores; i++) {
+			if (num_succ_notified_inst == gk_conf->num_lcores)
+				break;
+
+			if (!is_succ_notified[i]) {
+				int ret = notify_gk_instance(fib,
+					&gk_conf->instances[i]);
+				if (ret == 0) {
+					is_succ_notified[i] = true;
+					num_succ_notified_inst++;
+				}
+			}
+		}
+	}
+
+	if (num_succ_notified_inst != gk_conf->num_lcores) {
+		RTE_LOG(WARNING, GATEKEEPER,
+			"gk: %s successfully notifies only %d/%d instances\n",
+			__func__, num_succ_notified_inst, gk_conf->num_lcores);
+	}
+
+	/* Wait for all GK instances to synchronize. */
+	while (rte_atomic16_read(&fib->num_updated_instances)
+			< num_succ_notified_inst)
+		rte_pause();
+}
+
+/*
+ * This function is called by del_fib_entry_locked().
+ * Notice that, it doesn't stand on its own, and it's only
+ * a construct to make del_fib_entry_locked() readable.
+ */
+static struct gk_fib *
+check_fib_entry_for_del_locked(
```

Comment:
`check_fib_entry_for_del_locked()` is misnamed since this function is only removing a prefix from the LPM table. In fact, after the simplifications proposed in the other comments, this function may no longer have a reason to exist.